sdk/ml/azure-ai-ml/azure/ai/ml/_artifacts/_artifact_utilities.py

```diff
@@ -429,18 +441,243 @@ def _check_and_upload_env_build_context(
     show_progress: bool = True,
 ) -> Environment:
     if environment.path:
-        uploaded_artifact = _upload_to_datastore(
-            operations._operation_scope,
-            operations._datastore_operation,
-            environment.path,
+        uploaded_artifact = _upload_snapshot_to_datastore(
+            operation_scope=operations._operation_scope,
+            datastore_operation=operations._datastore_operation,
+            path=environment.path,
+            datastore_name=environment.datastore,
             asset_name=environment.name,
             asset_version=str(environment.version),
             asset_hash=environment._upload_hash,
-            sas_uri=sas_uri,
-            artifact_type=ErrorTarget.ENVIRONMENT,
-            datastore_name=environment.datastore,
             show_progress=show_progress,
+            sas_uri=sas_uri,
         )
         # TODO: Depending on decision trailing "/" needs to stay or not. EMS requires it to be present
         environment.build.path = uploaded_artifact.full_storage_path + "/"
     return environment
+
+
+@retry(
+    exceptions=HttpResponseError,
+    failure_msg="Artifact upload exceeded maximum retries. Try again.",
+    logger=module_logger,
+    max_attempts=MAX_ASSET_STORE_API_CALL_ATTEMPTS,
+)
+def get_temporary_data_reference(
+    operations: "DatastoreOperations",
+    asset_name: str,
+    asset_version: str,
+    request_headers: Dict[str, str],
+    asset_type: str = "codes",
+    ) -> Tuple[str, str]:
+    # create temporary data reference
+    temporary_data_reference_id = str(uuid.uuid4())
+
+    # get workspace credentials
+    resource_group_name = operations._resource_group_name
+    workspace_name = operations._workspace_name
+    workspace = operations._service_client.workspaces.get(
+        resource_group_name=resource_group_name,
+        workspace_name=workspace_name
+    )
+    workspace_id = workspace.workspace_id
+    workspace_location = workspace.location
+
+    # build asset id
+    asset_id = f"azureml://locations/{workspace_location}/workspaces/{workspace_id}/"\
+        f"{asset_type}/{asset_name}/versions/{asset_version}"
+
+    # build and send request
+    data = {
+        "assetId": asset_id,
+        "temporaryDataReferenceId": temporary_data_reference_id,
```

Comment:
From this spec, looks like they want to rename everything to pending upload id. Will there be a huge effort to rename it after this Pr ships? These generated codes can just be re-generated?

Reply:
What do you mean by "rename everything"? Change the name of the key in the data dictionary?